### DISCUSSION TITLE
Error in migration with enumeration

### DISCUSSION THREAD

**Discussion by adamaho** [SeaQL/sea-orm#1330] [Q&A]
Hi folks,

I seem to be struggling to get enumerations working in migrations.

Here is an example migration file:

```rust
use sea_orm_migration::prelude::*;

#[derive(DeriveMigrationName)]
pub struct Migration;

#[async_trait::async_trait]
impl MigrationTrait for Migration {
    async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager
            .create_table(
                Table::create()
                    .table(User::Table)
                    .if_not_exists()
                    .col(
                        ColumnDef::new(User::UserId)
                            .integer()
                            .not_null()
                            .auto_increment()
                            .primary_key(),
                    )
                    .col(ColumnDef::new(User::Email).string().not_null().unique_key())
                    .col(
                        ColumnDef::new(User::Role)
                            .enumeration(
                                RoleEnum::Enum,
                                [RoleEnum::Heron, RoleEnum::Admin, RoleEnum::User],
                            )
                            .not_null()
                            .default(String::from("admin")),
                    )
                    .col(
                        ColumnDef::new(User::CreatedAt)
                            .timestamp()
                            .not_null()
                            .extra(String::from("DEFAULT CURRENT_TIMESTAMP")),
                    )
                    .to_owned(),
            )
            .await
    }

    async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager
            .drop_table(Table::drop().table(User::Table).to_owned())
            .await
    }
}

#[derive(Iden)]
enum User {
    Table,
    UserId,
    Email,
    Role,
    CreatedAt,
}

#[derive(Iden)]
pub enum RoleEnum {
    #[iden = "Role"]
    Enum,
    #[iden = "heron"]
    Heron,
    #[iden = "admin"]
    Admin,
    #[iden = "user"]
    User,
}
```

When running the migrations I get the following error:

```
2022-12-22T18:29:11.300443Z  INFO sea_orm_migration::migrator: Applying migration 'm20221222_113332_create_users'
2022-12-22T18:29:11.305351Z  INFO sqlx::query: CREATE TABLE IF NOT …; rows affected: 0, rows returned: 0, elapsed: 1.847ms

CREATE TABLE IF NOT EXISTS "user" (
  "user_id" serial NOT NULL PRIMARY KEY,
  "email" varchar NOT NULL UNIQUE,
  "role" Role NOT NULL DEFAULT 'admin',
  "created_at" timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP
)
  
Execution Error: error returned from database: type "role" does not exist
```

From what I could see in the changelog for enumerations, this is what I am supposed to do to get this to work.

Anyone have any ideas on how to fix this?

Thanks!

**Comment by brianbruggeman** [comment]
I'm not super familiar with sea-orm, but it seems like you might be missing something with respect to the enumeration.  If you look at the documentation on enumeration here: https://www.sea-ql.org/SeaORM/docs/generate-entity/enumeration/

I think your enumeration may be missing at least one derive (DeriveActiveEnum) and a couple of other proc-macros:
* `#[sea_orm(rs_type = "String", db_type = "String(Some(1))")]`
* `#[sea_orm(string_value = "B")]`

**Reply by adamaho** [reply]
Thanks for the response! Interestingly enough, I thought the same. But I came across the changelog where this feature was updated and it said to do what I am doing above. Here is the reference I am referring to: https://github.com/SeaQL/sea-orm/blob/db6be517b1c292b9e5851d8b9b739f92edb9d03f/CHANGELOG.md#breaking-changes

It would be super helpful if the docs had an example of how to write this kind of migration :)

**Comment by tyt2y3** [comment]
If you are using Postgres and using DB-backed enum you have to create the enum with either https://docs.rs/sea-query/latest/sea_query/extension/postgres/struct.TypeCreateStatement.html
Or https://docs.rs/sea-orm/latest/sea_orm/schema/struct.Schema.html#method.create_enum_from_active_enum
